dist/amd64/birdy/bird-wrapper.sh

```diff
@@ -15,11 +25,171 @@ sed -i "s/BIRD_ROUTERID/${ROUTER_ID}/g" /etc/bird6.conf
+# that we still have the routes that we'd expect through that interface.
+try_update_nic_addr()
+{
+    addr=$1
+    intf=$2
+    tor_addr=$3
+    change_to_scope_link=$4
+
+    # Calculate the prefix length and IP network of the given address.
+    subnet_prefix_len=${addr#*/}
+    eval `ipcalc -n ${addr}`
+    subnet_network=$NETWORK
+
+    # Calculate the IP network of the subnet that $tor_addr is in (assuming the
+    # same prefix length as for the given NIC address).
+    eval `ipcalc -n ${tor_addr}/${subnet_prefix_len}`
+    tor_network=$NETWORK
+
+    # If the networks are the same...
+    if [ "$subnet_network" = "$tor_network" ]; then
+
+	if $change_to_scope_link; then
+	    # Delete the given address and re-add it with scope link.
+	    ip a d $addr dev $intf
+	    ip a a $addr dev $intf scope link
+	fi
+
+	# Ensure that the subnet route is present.  (This 'ip r a' will fail if
+	# the same route is already present.)
+	ip r a ${subnet_network}/${subnet_prefix_len} dev $intf || true
+
+	# Try to add a default route via the ToR.  (This 'ip r a' will fail if
+	# we already have a default route, e.g. via the other ToR.)
+	ip r a default via ${tor_addr} || true
+
+    fi
+}
+
+echo Invoked with: "$@"
+if [ "$1" = dual-tor ]; then
+
+    # Run with "dual-tor" argument: do setup for a dual ToR node, then run as
+    # the "early BGP" daemon until calico-node's BIRD can take over.
+
+    # There must be a following argument that is the URL of a shell include
+    # defining a `get_dual_tor_details` function.  Given one of the NIC-specific
+    # addresses for this node, this function must output variable settings for
+    # this node's stable address and AS number, and for the ToR addresses to
+    # peer with, for example like this:
+    #
+    # DUAL_TOR_STABLE_ADDRESS=172.31.20.3
+    # DUAL_TOR_AS_NUMBER=65002
+    # DUAL_TOR_PEERING_ADDRESS_1=172.31.21.100
+    # DUAL_TOR_PEERING_ADDRESS_2=172.31.22.100
+    #
+    # Note, the shell include code should not have any effects other than
+    # defining the required `get_dual_tor_details` function.
+    wget -O details.sh "$2"
+    . ./details.sh
+
+    echo "NIC-specific address is $ROUTER_ID"
+    eval `get_dual_tor_details $ROUTER_ID`
+    echo "Stable (loopback) address is $DUAL_TOR_STABLE_ADDRESS"
+    echo "AS number is $DUAL_TOR_AS_NUMBER"
+    echo "ToR addresses are $DUAL_TOR_PEERING_ADDRESS_1 and $DUAL_TOR_PEERING_ADDRESS_2"
 
-# Loop sleeping
-while true; do
-    sleep 60
-done
+    # Configure the stable address.
+    ip a a ${DUAL_TOR_STABLE_ADDRESS}/32 dev lo
+
+    # Generate BIRD peering config.
+    cat >/etc/bird/peers.conf <<EOF
+template bgp tors {
+  description "Connection to ToR";
+  local as $DUAL_TOR_AS_NUMBER;
+  direct;
+  gateway recursive;
+  import all;
+  export all;
+  add paths on;
+  connect delay time 2;
+  connect retry time 5;
+  error wait time 5,30;
+  next hop self;
+  bfd on;
```

Comment:
Surely BFD needs to be configurable?  they may not want that.

Reply:
I don't think it can do any harm, can it?